tests/main/snapd-snap/task.yaml

```diff
@@ -1,27 +1,143 @@
 summary: Ensure snapd builds as a snap
 
-# This is what the snapcraft builder on launchpad uses.
-systems: [ubuntu-16.04-64]
+# TODO: we don't need to build the snapd snap in this test on as many systems
+# when we have a way to dynamically download the snapd snap that was built for a
+# particular PR from github actions, in which case we can just download that and
+# install it here to test the things that we need to test about the snapd snap
+
+# effectively we can only build the snap using LXD + snapcraft on ubuntu systems
+# due to various bugs with other distros and snapcraft and/or LXD
+systems:
+    # snapcraft is a classic snap, can't run on ubuntu-core
+    - -ubuntu-core*
+    # LXD doesn't work properly on amazon linux, fedora 31, opensuse, or debian
+    # 9 to build the snap, it complains about fuse.snapfuse getting EPERM when
+    # snaps are installed inside the LXD container that snapcraft creates
+    - -amazon-linux*
+    - -debian-*
+    - -fedora-*
+    - -opensuse-*
+    # additionally on fedora 32 and 33, and arch linux which are all cgroupsv2
+    # now, LXD does not work either, with the service failing to start up at all
+    # when we install LXD
+    - -arch-*
+    # snapcraft does not support running on centos 7 due to errors with the
+    # locale, see https://bugs.launchpad.net/snapcraft/+bug/1922140
+    # a separate error occurs on centos 8, see https://bugs.launchpad.net/snapcraft/+bug/1922981
+    - -centos-*
 
 # Start early as it takes a long time.
 priority: 100
 
+environment:
+    # the destructive build mode runs only on xenial, but we also run lxd on
+    # xenial to make sure it builds there too
+    SNAPCRAFT_BUILD_ENVIRONMENT/destructive: host
+    SNAPCRAFT_BUILD_ENVIRONMENT/lxd: lxd
+
 prepare: |
+    # we need to use tests.cleanup from the same dir that we run snapcraft in 
+    # such that we don't need to change directories from the cleanup restore 
+    # script
+    # shellcheck disable=SC2164
+    cd "$PROJECT_PATH"
+    tests.cleanup prepare
+
+    # only run the destructive mode variant on xenial
+    if ! os.query is-xenial && [ "$SNAPCRAFT_BUILD_ENVIRONMENT" = "host" ]; then
+        echo "Skipping destructive mode build on non-xenial"
+        exit 0
+    fi
+
+    if os.query is-fedora || os.query is-arch-linux || os.query is-centos; then
+        # in order to install classic snaps, we need to setup the /snap 
+        # symlink on these distros
+        ln -sf /var/lib/snapd/snap /snap
```

Comment:
In other tests we do:
            SNAP_MOUNT_DIR="$(os.paths snap-mount-dir)"
            ln -sf "$SNAP_MOUNT_DIR" /snap